### ISSUE TITLE
Proactively test reflow

### ISSUE THREAD

**Issue by vanrein** [vanrein/perpetuum#15]
* Limit the catch-all in GenMod:transmap/1 so that one placebit more wouldn't fit
  * Make a new testrun:

     1. Incrementally apply reflow_transmap()
     2. Invoke GenMap:transmap() with incremental numbers
     3. Compare these transmaps
     4. This crashes when bit counts are off
     5. Make 10 iterations, more than the compiler generates
     6. Perhaps: Make a syncrun() against each outcome

**Comment by vanrein** [comment]
NP for kxover_client or kxover_server, but a bug in the much simpler traffic_light_nightly.


{'mappings-differ',dawn,

{17498005798264095394980017816940970922825355447145699491406164851279623993595007385788105416184430592,
{17498005798264095394980017816940970922825355447145699491406164851279623993595007385788105416184430592,
0,
0,
45424647598781632625707920442389838554292749850318596674233162151666353298854802420341240474156899228959002286592435789810634162438144},
45427420268475430659332737993000274648099685910909681277584228978307864087743040465011528343413674662867939660574467705485966543159296}},

{'mappings-differ',stop,

{2596148429267413814265248164610048,
{2596148429267413814265248164610048,
6739986666787659948666753771754907668409286105635143120275902562304,
6739986666787659948666753771754907668409286105635143120275902562304,
45424647598781632625707920442389838554292749850318596674233162151666353298854802420341240474156899228959002286592435789810634162438144},
45427420268475430659332737993000274648099685910909681277584228978307864087743040465011528343413674662867939660574467705485966543159296}},

[...and so on... last 6 entries of 11, so what's all generated by reflow_transmap) ]



The difference lies in the last word, or the TransSentinel, as is clear from:

```
        transmap  :: #{
                atom() =>               % TransName
                {                       % TransInfo
                        integer(),      % MarkingAddend
                        integer(),      % MarkingSubber
                        integer()       % TransSentinel
                }
        }
```

**dawn**

```
>>> a=45424647598781632625707920442389838554292749850318596674233162151666353298854802420341240474156899228959002286592435789810634162438144
>>> b=45427420268475430659332737993000274648099685910909681277584228978307864087743040465011528343413674662867939660574467705485966543159296
>>> '%x'%a
'fffc00000001ffffffffffffffff00000000000000000000000000020000000000000000000000000004000000000000000000000000000'
>>> '%x'%b
'ffffffffffffffffffffffffffff00000000000000000000000000020000000000000000000000000004000000000000000000000000000'
```

**stop**

```
>>> c=45424647598781632625707920442389838554292749850318596674233162151666353298854802420341240474156899228959002286592435789810634162438144
>>> d=45427420268475430659332737993000274648099685910909681277584228978307864087743040465011528343413674662867939660574467705485966543159296
>>> '%x'%c
'fffc00000001ffffffffffffffff00000000000000000000000000020000000000000000000000000004000000000000000000000000000'
>>> '%x'%d
'ffffffffffffffffffffffffffff00000000000000000000000000020000000000000000000000000004000000000000000000000000000'
```

Initial values are:

```
initial_placebits() -> 13.
```

For this value, the transition maps for stop and dawn are:

```
'stop' => { 16384, 268435456, 72055395148898304 },
'dawn' => { 4398046511104, 0, 72055395148898304 },
```

The same TransSentinel 72055395148898304, or fffe0008002000.

Clearly, the error is in the inhibiting place, night_service, which is indeed the last in the places list:

```
places() -> [ 'green', 'red', 'yellow', 'night_service' ].
```

Also, the transitions are precisely those that ought to be inhibited.


Inhibitors in a transmap should be all ones, so
  - the wrong value is fffc00000001ffffffffffffffff (from a/c)
  - the right value is ffffffffffffffffffffffffffff (from b/d)

An earlier sample:

```
                   dawn =>
                       {784637716923335095479473677900958302012794430558004314112,
                        0,
                        7237005577332262213580867704581326693132172498517173409397235297162966859776},
                   stop =>
                       {9223372036854775808,
                        85070591730234615865843651857942052864,
                        7237005577332262213580867704581326693132172498517173409397235297162966859776},
```

These present the TransSentinel as the same value

```
ffffffffffffffff00000000000000020000000000000004000000000000000
```

And that is still correct.  It occurs 4x, twice (dawn + stop) for each side of the comparison.


The problem begins when reflowing dynamically, restarting at the smallest PlaceBits and TransMap.
This is a separate clause in the generated Erlang module.  But the correct values are listed under
"Module produces"... the code there is:

```
transmap(    N ) when (((N+1)*4+1) div 64) < (((N+2)*4+1) div 64) ->
                gen_perpetuum:reflow_transmap( transmap( 13 ),4,13,N ).
% Note: reflow_transmap( TransMap,NumPlaces,OldPlacebits,NewPlacebits )
```

Yes, that does look good...  and so does the last TransSentinel value in the Module,

```
>>> e=2462625387274654950767440006227893160541872039269379026238376672976002436389220904551283407717126119502087303725056
>>> '%x\n'%e
'ffffffffffffffffffffffff00000000000000000000000200000000000000000000000400000000000000000000000\n'
```

It might be something else that switches... like the Sentinel:

```
sentinel(    N ) -> gen_perpetuum:reflow_sentinel( sentinel( 13 ),4,13,N ).
% Note: reflow_sentinel( Vec,NumPlaces,OldPlacebits,NewPlacebits )
```

But that one also looks good.  It would have impacted both, anyway.

*Hmmm...*

**Comment by vanrein** [comment]
The problem *does* occur when switching the approach from precompiled output to a jump up from the smallest value!  That makes sense :)

```
-Reflowed Sentinel 72055395148898304 to 45424647598781632625707920442389838554292749850318596674233162151666353298854802420341240474156899228959002286592435789810634162438144
```
vs.
```
+Reflowed Sentinel 2462625387274654950767440006227893160541872039269379026238376672976002436389220904551283407717126119502087303725056 to 45427420268475430659332737993000274648099685910909681277584228978307864087743040465011528343413674662867939660574467705485966543159296
```

The former is
```
fffc00000001ffffffffffffffff00000000000000000000000000020000000000000000000000000004000000000000000000000000000
```
the latter is
```
ffffffffffffffffffffffffffff00000000000000000000000000020000000000000000000000000004000000000000000000000000000
```

With `initial_placebits() -> 13` we have
```
>>> '%x'%(0xfffe0008002000 >> 3*14)
'3fff'
>>> '%x'%(((1<<14)-1)<< 3*14)
'fffc0000000000'
>>> '%x'%72055395148898304
'fffe0008002000'
```
Note that `fffe` includes a sentinel bit from the 3rd place that is otherwise filled with zeroes.  So this correctly differs from `fffc` in the TransSentinel alone.

Our problem is located in `gen_perpetuum:reflow_sentinel()`, and deals with the large  step up.

**Comment by vanrein** [comment]
The sentinel is shifted up and filled with zeroes, rather than with repeated bits as we had them.  This is why the original 13/14 set bits keep showing up as `fffe` and so on.  This is the intended behaviour for `NewVec`,

```
                ReguardExtraBits = ( -1 bsl (NewPlacebits-OldPlacebits) ) bxor -1,
                NewVec = if
                        (Vec band (1 bsl BitPos)) /= 0 -> ReguardExtraBits bsl BitPos;
                        true -> 0
                end,
```
In the last iteration, we have NewVec as
```
400000001fffffffffffffffe00000000000000000000000000000000000000000000000000000000000000000000000000000000000
```
which is indeed part of the wrong output,
```
fffc00000001ffffffffffffffff00000000000000000000000000020000000000000000000000000004000000000000000000000000000
```
where it shows that it is quite simply too short.

In this position in the test, we have

  * `OldPlacebits = 13`
  * `NewPlacebits = 110`
 
So the expected value is
```
>>> (-1 << (110-13) ) ^ -1
158456325028528675187087900671L

1> NewPlacebits = 110.
2> OldPlacebits = 13.
3> ReguardExtraBits = ( -1 bsl (NewPlacebits-OldPlacebits) ) bxor -1.
158456325046975419260797452287

>>> '%x'%158456325028528675187087900671
'1ffffffffffffffffffffffff'
>>> '%x'%158456325046975419260797452287
'200000000ffffffffffffffff'
```

We have pinned down our bug.  Now to smash it.

Oh, and `ReguardExtraBits` can also move in for efficiency.

**Comment by vanrein** [comment]
This is starting to look like a bug in Erlang!  Given the large integers however, I would not expect it to be a security threat in any practical setting.

```
3> ReguardExtraBits = ( -1 bsl (NewPlacebits-OldPlacebits) ) bxor -1.
158456325046975419260797452287
4> -1 bsl (NewPlacebits-OldPlacebits).
-158456325028528675187087900672
5> (-1 bsl (NewPlacebits-OldPlacebits)) bxor -1.
158456325046975419260797452287
```
with hexadecimal notation
```
>>> '%x'%158456325046975419260797452287
'200000000ffffffffffffffff'
>>> '%x'% -158456325028528675187087900672
'-2000000000000000000000000'
>>> '%x'% 158456325046975419260797452287
'200000000ffffffffffffffff'
>>> len ('ffffffffffffffff')*4
64
```

This looks suspiciously like the -1 argument to `bxor` was not extended, or something similar.

This is definitely leaning on a rugid corner of the math, I am aware of that, but I can find no reason why the integer format should reveal details about internal word sizes.  If it works for 10 bits, it should also work for 110 bits, at least that is the logic I read from the arbitrary-sized integers in Erlang.

**Comment by vanrein** [comment]
Yup, Erlang bug...
https://bugs.erlang.org/browse/ERL-450

Fix Version/s: 20.1, 21.0

I will not check for the version -- you can run CTest and make sure for yourself that you are fine.  If you believe you may need more storage capacity than precomputed by the perpetuum compiler, then you would not like your system to break down just at the point it gets there.  So run your checks to assure yourself that your system is up to date and if not, choose whether you are going to take the risk of these incredibly high numbers happening to you.

The mathematics involved:
  * NumPlaces is the number of places in your Petri Net
  * Placebits is the number of bits you need to capture the highest number of tokens in unsigned binary

Then, *1 + NumPlaces * (PlaceBits + 1)* must not exceed 64 bits, though I would prefer not going over 63 bits because this is a sign-related problem and integers in Erlang are stored in 2's complement.

**Comment by vanrein** [comment]
Solved in 36994537395a6f344d6a0fa1f52fe236b8b5250f or, more popularly put, version 1.0.0.
